File: Backend/src/main/java/com/sencarmarket/module/garage/service/GarageService.java
package com.sencarmarket.module.garage.service;

import com.sencarmarket.module.commun.dto.PaginatedResponse;
import com.sencarmarket.module.garage.dto.*;

import java.util.List;
import java.util.UUID;

/**
 * Interface pour le service garage
 * Implémente le principe de DIP (Dependency Inversion Principle)
 */
public interface GarageService {

    // ========== GARAGE ==========

    /**
     * Créer un nouveau garage
     */
    GarageResponse createGarage(CreateGarageRequest request, UUID proprietaireId);

    /**
     * Mettre à jour un garage
     */
    GarageResponse updateGarage(UUID id, CreateGarageRequest request);

    /**
     * Supprimer un garage
     */
    void deleteGarage(UUID id);

    /**
     * Obtenir un garage par ID
     */
    GarageResponse getGarageById(UUID id);

    /**
     * Obtenir tous les garages avec pagination
     */
    PaginatedResponse<GarageResponse> getAllGarages(int page, int size);

    /**
     * Obtenir les garages actifs avec pagination
     */
    PaginatedResponse<GarageResponse> getActiveGarages(int page, int size);

    /**
     * Obtenir les garages en attente de validation
     */
    PaginatedResponse<GarageResponse> getGaragesEnAttente(int page, int size);

    /**
     * Obtenir les garages d'un propriétaire
     */
    List<GarageResponse> getGaragesByProprietaire(UUID proprietaireId);

    /**
     * Rechercher des garages par localisation
     */
    List<GarageResponse> searchByLocalisation(String ville);

    /**
     * Rechercher des garages par proximité
     */
    List<GarageResponse> searchByProximity(Double latitude, Double longitude, Double rayonKm);

    /**
     * Rechercher des garages
     */
    List<GarageResponse> searchGarages(String query);

    /**
     * Valider un garage (admin)
     */
    GarageResponse validerGarage(UUID id, ValidationGarageRequest request);

    /**
     * Mettre à jour le logo
     */
    GarageResponse updateLogo(UUID id, String logoUrl);

    // ========== SERVICE GARAGE ==========

    /**
     * Créer un service
     */
    ServiceGarageResponse createService(CreateServiceGarageRequest request);

    /**
     * Mettre à jour un service
     */
    ServiceGarageResponse updateService(UUID id, CreateServiceGarageRequest request);

    /**
     * Supprimer un service
     */
    void deleteService(UUID id);

    /**
     * Obtenir un service par ID
     */
    ServiceGarageResponse getServiceById(UUID id);

    /**
     * Obtenir tous les services
     */
    List<ServiceGarageResponse> getAllServices();

    /**
     * Obtenir les services par catégorie
     */
    List<ServiceGarageResponse> getServicesByCategorie(String categorie);

    // ========== GARAGE-SERVICE (ASSOCIATION) ==========

    /**
     * Associer un service à un garage
     */
    GarageServiceResponse associateService(AssociateServiceRequest request);

    /**
     * Mettre à jour l'association
     */
    GarageServiceResponse updateAssociation(UUID id, AssociateServiceRequest request);

    /**
     * Supprimer l'association
     */
    void disassociateService(UUID garageId, UUID serviceId);

    /**
     * Obtenir les services d'un garage
     */
    List<GarageServiceResponse> getServicesByGarage(UUID garageId);

    /**
     * Obtenir les garages proposant un service
     */
    List<GarageServiceResponse> getGaragesByService(UUID serviceId);
}

File: Frontend/src/app/features/garages/services/garage.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { Garage, ServiceGarage, GarageServiceAssociation } from '../models/garage.model';

@Injectable({
  providedIn: 'root'
})
export class GarageService {
  private apiUrl = `${environment.apiUrl}/garages`;

  constructor(private http: HttpClient) {}

  getAllGarages(page = 0, size = 10): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(this.apiUrl, { params });
  }

  getActiveGarages(page = 0, size = 10): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/actifs`, { params });
  }

  getGarageById(id: string): Observable<Garage> {
    return this.http.get<Garage>(`${this.apiUrl}/${id}`);
  }

  getGaragesByVille(ville: string): Observable<Garage[]> {
    return this.http.get<Garage[]>(`${this.apiUrl}/search/ville`, { params: { ville } });
  }

  searchGarages(q: string): Observable<Garage[]> {
    return this.http.get<Garage[]>(`${this.apiUrl}/search`, { params: { q } });
  }

  getServicesByGarage(garageId: string): Observable<GarageServiceAssociation[]> {
    return this.http.get<GarageServiceAssociation[]>(`${this.apiUrl}/${garageId}/services`);
  }

  createGarage(garage: any): Observable<Garage> {
    return this.http.post<Garage>(this.apiUrl, garage);
  }

  updateGarage(id: string, garage: any): Observable<Garage> {
    return this.http.put<Garage>(`${this.apiUrl}/${id}`, garage);
  }

  deleteGarage(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getGaragesEnAttente(page = 0, size = 10): Observable<any> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/en-attente`, { params });
  }

  getGaragesByProprietaire(proprietaireId: string): Observable<Garage[]> {
    return this.http.get<Garage[]>(`${this.apiUrl}/proprietaire/${proprietaireId}`);
  }

  searchByProximity(latitude: number, longitude: number, rayonKm = 10): Observable<Garage[]> {
    const params = new HttpParams()
      .set('latitude', latitude.toString())
      .set('longitude', longitude.toString())
      .set('rayonKm', rayonKm.toString());
    return this.http.get<Garage[]>(`${this.apiUrl}/search/proximity`, { params });
  }

  validerGarage(id: string, request: any): Observable<Garage> {
    return this.http.post<Garage>(`${this.apiUrl}/${id}/validate`, request);
  }

  updateLogo(id: string, logoUrl: string): Observable<Garage> {
    return this.http.put<Garage>(`${this.apiUrl}/${id}/logo`, { logoUrl });
  }

  createService(request: any): Observable<ServiceGarage> {
    return this.http.post<ServiceGarage>(`${this.apiUrl}/services`, request);
  }

  getAllServices(): Observable<ServiceGarage[]> {
    return this.http.get<ServiceGarage[]>(`${this.apiUrl}/services`);
  }

  getServiceById(id: string): Observable<ServiceGarage> {
    return this.http.get<ServiceGarage>(`${this.apiUrl}/services/${id}`);
  }

  associateService(garageId: string, request: any): Observable<GarageServiceAssociation> {
    return this.http.post<GarageServiceAssociation>(`${this.apiUrl}/${garageId}/services`, request);
  }

  disassociateService(garageId: string, serviceId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${garageId}/services/${serviceId}`);
  }
}
